nmstatectl/nmstatectl.py

```diff
@@ -56,18 +57,26 @@ def setup_subcommand_set(subparsers):
 def show(args):
     state = netinfo.show()
     if args.yaml:
-        print(yaml.dump(state, default_flow_style=False))
+        if sys.stdout.isatty():
+            print(yaml.dump(state, default_flow_style=False))
+        else:
+            print(yaml.dump(state))
     else:
-        print(json.dumps(state, indent=4, sort_keys=True,
-                         separators=(',', ': ')))
+        if sys.stdout.isatty():
+            print(json.dumps(state, indent=4, sort_keys=True,
+                            separators=(',', ': ')))
+        else:
+            print(json.dumps(state))
 
 
 def apply(args):
-    with open(args.file) as f:
-        statedata = f.read()
-    try:
+    with open(args.file) as statefile:
+        statedata = statefile.read()
+
+    # JSON dictionaries start with a curly brace
+    if statedata[0] == '{':
```

Comment:
If the file is empty it will raise in an non descriptive manner.
You need to handle that edge case (len(statedata) == 0).

Reply:
Good catch, should it be an error? And how would you like it to be reported? I guess non-zero returncode (for example 1) and a message to stderr "ERROR: statefile is empty". Or should be considert this as nothing needs to be done and just exit?

Reply:
Like all the other errors... raise. And better raise a dedicated error (EmptyFileError).
When raising, the argument can be the file path:
'''
class EmptyFileError(Exception):
    pass

raise EmptyFileError(file)
'''

Reply:
Are you sure? This is in the tool and not the library, IMHO an exception is not a good user error message in this case. And is it really an error? If it is getting data from stdin and the other tool that pipes its output to nmstatectl does not have a state to apply, getting no data sounds like a valid scenario.

Reply:
You will have to catch all exceptions then: (missing file, no permission to open, etc..)
Then log it to stderr and so.

I guess you could solve it by raising at the error location, but translating it at the main entry point (from an exception to a stderr print/log in a generic fashion).
BTW, it is improper to exit the program in multiple places.

Reply:
Perfect is the enemy of good. Ideally, all exceptions would be caught, however, if we just want exceptions to be shown to the user, `len(statedata) == 0` does not need to be caught... IMHO it is still better to not raise an exception when it is meant for the end user even when there are other exceptions that could be raised.

Reply:
There are several errors we could expect to see: Missing file, empty file, corrupted json/yaml file and so on. Which one you prefer the user to see as an exception and which one you want the user to just get an stderr output? What is the rule here?
We do not need perfect here, we just need consistency.
As an example, you could say that a bug in the code should throw an exception while an expected error in the input should just print the error ad exit with a non zero code.
The easiest path I guess is to throw.
An upgrade to that is to serialize the error using stderr for specific exceptions.
Lets just be consistent as much as we can.